### ISSUE TITLE
package matplotlib 1.2.1 download broken

### ISSUE THREAD

**Issue by skaffen-amtiskaw** [galaxyproject/tools-iuc#519]
```
matplotlib	1.2.1	package	Error	  File "/opt/galaxy/lib/tool_shed/galaxy_install/install_manager.py", line 128, in install_and_build_package_via_fabric
    tool_dependency = self.install_and_build_package( install_environment, tool_dependency, actions_dict )
  File "/opt/galaxy/lib/tool_shed/galaxy_install/install_manager.py", line 86, in install_and_build_package
    initial_download=True )
  File "/opt/galaxy/lib/tool_shed/galaxy_install/tool_dependencies/recipe/recipe_manager.py", line 31, in execute_step
    initial_download=initial_download )
  File "/opt/galaxy/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py", line 694, in execute_step
    dir = self.url_download( work_dir, downloaded_filename, url, extract=True, checksums=checksums )
  File "/opt/galaxy/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py", line 187, in url_download
    raise Exception( err_msg )

Error downloading from URL
https://github.com/matplotlib/matplotlib/archive/v1.2.1.tar.gz#md5#47ee56f51200b8760b98d97ee86f57c6:
HTTP Error 400: Bad Request
```
url: https://github.com/matplotlib/matplotlib/archive/v1.2.1.tar.gz#md5#47ee56f51200b8760b98d97ee86f57c6:

**Comment by martenson** [comment]
that URL works fine for me, maybe the machine that is trying to download it has connection trouble?

**Comment by skaffen-amtiskaw** [comment]
Yeah and it works fine from a wget on the same machine galaxy is running on. I don't get it. Is there a way to manually place the tar file so that galaxy can just see that its there and get on with the install?

**Comment by martenson** [comment]
@skaffen-amtiskaw Not as a part of TS installation I believe. Is the semicolon in the end of the url legit?

**Comment by skaffen-amtiskaw** [comment]
I guess so? It's just the error log from the galaxy toolshed install.

**Comment by yhoogstrate** [comment]
The galaxy code (https://github.com/galaxyproject/galaxy/blob/dev/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py#L169) works:

```
import urllib2
import time
download_url = "https://github.com/matplotlib/matplotlib/archive/v1.2.1.tar.gz#md5#47ee56f51200b8760b98d97ee86f57c6"
file_path = "/tmp/a"
dst = open( file_path, 'wb' )
while True:
    chunk = src.read(1000 )
    if chunk:
            dst.write( chunk )
    else:
            break
```

@skaffen-amtiskaw did the error occur just once or is it permanent? Can you run the piece of code above, just to make sure your ulrlib2 has no issues?

**Comment by mvdbeek** [comment]
I also encountered this a few times (with other downloads). It seems wget/curl is a bit more reliable than urllib.

**Comment by skaffen-amtiskaw** [comment]
@yhoogstrate The error occurs every time. 
pasta:
Traceback (most recent call last):
  File "check.py", line 5, in <module>
    src = urllib2.urlopen( download_url )
  File "/usr/lib64/python2.6/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib64/python2.6/urllib2.py", line 397, in open
    response = meth(req, response)
  File "/usr/lib64/python2.6/urllib2.py", line 510, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib64/python2.6/urllib2.py", line 435, in error
    return self._call_chain(*args)
  File "/usr/lib64/python2.6/urllib2.py", line 369, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.6/urllib2.py", line 518, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
urllib2.HTTPError: HTTP Error 400: Bad Request

**Comment by yhoogstrate** [comment]
Nice... I can reproduce this error using python2.6.

The python 2.6 version of python can't cope with hashtags which should be stripped of somewhere.

@erasche Do you want to keep the hashtags supported?

**Comment by bgruening** [comment]
@yhoogstrate this can now moved to the attribute `sha256=` and can be removed.

**Comment by yhoogstrate** [comment]
@bgruening Absolutely, @erasche has fixed this in #213.

**Comment by yhoogstrate** [comment]
@skaffen-amtiskaw  the pull request that is supposed to solve this issue has been merged yesterday and I have just updated them in the tool shed. Can you confirm whether this solves your problem?

If this indeed solves it for you, could you close the issue?

Thanks again for reporting.
